CSP: "Missing CSP nonce" on error page

### Describe the bug

When a POST request throws an exception and is redirected to the configured custom error page, Primefaces fails to get the current CSP and throws a `CspException`, preventing the custom error page from being displayed. Instead, Tomcat renders its default error page for the original exception and logs the following

```
18-Nov-2024 14:40:45.195 SCHWERWIEGEND [http-nio-8080-exec-5] org.apache.catalina.core.StandardHostValve.custom Exception Processing [ErrorPage[errorCode=0, location=/error.xhtml]]
	jakarta.servlet.ServletException: Missing CSP nonce
		at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:709)
		at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:449)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
		at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:632)
		at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:410)
		at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:330)
		at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:267)
		at org.apache.catalina.core.StandardHostValve.custom(StandardHostValve.java:362)
		at org.apache.catalina.core.StandardHostValve.status(StandardHostValve.java:222)
		at org.apache.catalina.core.StandardHostValve.throwable(StandardHostValve.java:308)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:149)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:663)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:397)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:905)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
		at java.base/java.lang.Thread.run(Thread.java:1583)
	Caused by: org.primefaces.csp.CspException: Missing CSP nonce
		at org.primefaces.csp.CspState.validate(CspState.java:92)
		at org.primefaces.csp.CspState.getNonce(CspState.java:71)
		at org.primefaces.csp.CspPhaseListener.initCsp(CspPhaseListener.java:94)
		at org.primefaces.csp.CspPhaseListener.beforePhase(CspPhaseListener.java:66)
		at com.sun.faces.lifecycle.Phase.handleBeforePhase(Phase.java:179)
		at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:70)
		at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:178)
		at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:692)
		... 24 more
```

### Reproducer

[demo.zip](https://github.com/user-attachments/files/17801071/demo.zip)

/webapp/index.xhtml
```xml
<!DOCTYPE html>
<html xmlns:h="jakarta.faces.html" lang="en">
<h:head>
    <title>Index</title>
</h:head>
<h:body>
    <h:form>
        <h:commandButton action="#{indexBean.action()}" value="Button"/>
    </h:form>
</h:body>
</html>
```

/webapp/error.xhtml
```xml
<!DOCTYPE html>
<html xmlns:h="jakarta.faces.html" lang="en">
<h:head>
    <title>Error</title>
</h:head>
<h:body>
    An error occurred.
</h:body>
</html>
```

/java/org/example/demo1/IndexBean.java
```java
package org.example.demo1;

import jakarta.enterprise.context.RequestScoped;
import jakarta.inject.Named;

@Named
@RequestScoped
public class IndexBean {

    public void action() {
        throw new RuntimeException("Something has gone wrong.");
    }
}
```

/webapp/WEB-INF/web.xml
```xml
<?xml version="1.0" encoding="UTF-8"?>
<web-app version="6.0" xmlns="https://jakarta.ee/xml/ns/jakartaee"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/web-app_6_0.xsd">

    <context-param>
        <param-name>primefaces.CSP</param-name>
        <param-value>true</param-value>
    </context-param>

    <servlet>
        <servlet-name>Faces Servlet</servlet-name>
        <servlet-class>jakarta.faces.webapp.FacesServlet</servlet-class>
        <load-on-startup>1</load-on-startup>
    </servlet>

    <servlet-mapping>
        <servlet-name>Faces Servlet</servlet-name>
        <url-pattern>*.xhtml</url-pattern>
    </servlet-mapping>

    <error-page>
        <location>/error.xhtml</location>
    </error-page>
</web-app>
```

### Expected behavior

Primefaces should be aware of error requests and not handle them the same as all other postback requests. A new CspState should be created and the custom error page should be rendered.

### PrimeFaces edition

Community

### PrimeFaces version

14.0.7

### Theme

_No response_

### JSF implementation

Mojarra

### JSF version

4.0.4

### Java version

21.0.2

### Browser(s)

_No response_

--- a/primefaces/src/main/java/org/primefaces/csp/CspPhaseListener.java
+++ b/primefaces/src/main/java/org/primefaces/csp/CspPhaseListener.java
@@ -29,6 +29,7 @@ import org.primefaces.context.PrimeFacesContext;
 import org.primefaces.util.LangUtils;
 import org.primefaces.util.Lazy;
 
+import javax.faces.application.ProjectStage;
 import javax.faces.context.ExternalContext;
 import javax.faces.context.FacesContext;
 import javax.faces.event.PhaseEvent;
@@ -96,7 +97,10 @@ public class CspPhaseListener implements PhaseListener {
             externalContext.addResponseHeader("Content-Security-Policy", policy);
         }
 
-        String init = "if(window.PrimeFaces){PrimeFaces.csp.init('" + Encode.forJavaScript(state.getNonce()) + "');};";
+        String init = "if(window.PrimeFaces){PrimeFaces.csp.init('" + Encode.forJavaScript(state.getNonce()) + "');}";
+        if (context.isProjectStage(ProjectStage.Development)) {
+            init += "else{console.log('CSP active but PrimeFaces not included in current view!');}";
+        }
         PrimeFaces.current().executeInitScript(init);
     }
 

--- a/primefaces/src/main/java/org/primefaces/csp/CspState.java
+++ b/primefaces/src/main/java/org/primefaces/csp/CspState.java
@@ -34,6 +34,7 @@ import java.util.Objects;
 import java.util.UUID;
 
 import javax.faces.context.FacesContext;
+import javax.servlet.http.HttpServletRequest;
 
 public class CspState {
 
@@ -84,6 +85,12 @@ public class CspState {
                 // always prefer nonceViewState over nonceRequest
                 nonce = LangUtils.isNotBlank(nonceViewState) ? nonceViewState : nonceRequest;
 
+                // in case of a forward, we might create a new nonce here
+                // but it also means that the request had NO CSP request param... is this valid?
+                if (LangUtils.isBlank(nonce) && isForward(context)) {
+                    nonce = Base64.getEncoder().encodeToString(UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8));
+                }
+
                 validate(nonce);
             }
             // otherwise create a new nonce
@@ -96,6 +103,14 @@ public class CspState {
         }
 
         return nonce;
+    }
+
+    protected boolean isForward(FacesContext context) {
+        Object request = context.getExternalContext().getRequest();
+        if (request instanceof HttpServletRequest) {
+            return ((HttpServletRequest) request).getAttribute("javax.servlet.forward.request_uri") != null;
+        }
+        return false;
     }
 
     /**

